token/core/fabtoken/wallet.go

```diff
@@ -7,251 +7,61 @@ SPDX-License-Identifier: Apache-2.0
-		Deserializer:           Deserializer,
-		OwnerWalletsRegistry:   OwnerWalletRegistry,
-		IssuerWalletsRegistry:  IssuerWalletRegistry,
-		AuditorWalletsRegistry: AuditorWalletRegistry,
-	}
 }
 
-func (s *WalletService) RegisterOwnerWallet(id string, path string) error {
-	return s.OwnerWalletsRegistry.RegisterIdentity(id, path)
+func NewWalletFactory(identityProvider driver.IdentityProvider, tokenVault TokenVault) *WalletFactory {
+	return &WalletFactory{IdentityProvider: identityProvider, TokenVault: tokenVault}
 }
 
-func (s *WalletService) RegisterIssuerWallet(id string, path string) error {
-	return s.IssuerWalletsRegistry.RegisterIdentity(id, path)
-}
-
-func (s *WalletService) RegisterRecipientIdentity(data *driver.RecipientData) error {
-	logger.Debugf("register recipient identity [%s] with audit info [%s]", data.Identity.String(), hash.Hashable(data.AuditInfo).String())
-	// recognize identity and register it
-
-	// match identity and audit info
-	err := s.Deserializer.Match(data.Identity, data.AuditInfo)
+func (w *WalletFactory) NewWallet(role driver.IdentityRole, walletRegistry common.WalletRegistry, info driver.IdentityInfo, id string) (driver.Wallet, error) {
+	idInfoIdentity, _, err := info.Get()
 	if err != nil {
-		return errors.Wrapf(err, "failed to match identity to audit infor for [%s:%s]", data.Identity, hash.Hashable(data.AuditInfo))
-	}
-
-	// register verifier and audit info
-	v, err := s.Deserializer.GetOwnerVerifier(data.Identity)
-	if err != nil {
-		return errors.Wrapf(err, "failed getting verifier for [%s]", data.Identity)
-	}
-	if err := s.IdentityProvider.RegisterVerifier(data.Identity, v); err != nil {
-		return errors.Wrapf(err, "failed registering verifier for [%s]", data.Identity)
-	}
-	if err := s.IdentityProvider.RegisterAuditInfo(data.Identity, data.AuditInfo); err != nil {
-		return errors.Wrapf(err, "failed registering audit info for [%s]", data.Identity)
+		return nil, errors.WithMessagef(err, "failed to get owner wallet identity for [%s]", id)
 	}
 
-	return nil
-}
-
-func (s *WalletService) GetAuditInfo(id view.Identity) ([]byte, error) {
-	return s.IdentityProvider.GetAuditInfo(id)
-}
-
-func (s *WalletService) GetEnrollmentID(auditInfo []byte) (string, error) {
-	return s.IdentityProvider.GetEnrollmentID(auditInfo)
-}
-
-func (s *WalletService) GetRevocationHandler(auditInfo []byte) (string, error) {
-	return s.IdentityProvider.GetRevocationHandler(auditInfo)
-}
-
-func (s *WalletService) Wallet(identity view.Identity) driver.Wallet {
-	w, _ := s.OwnerWalletByIdentity(identity)
-	if w != nil {
-		return w
+	var newWallet driver.Wallet
+	switch role {
+	case driver.OwnerRole:
+		newWallet = newOwnerWallet(w.IdentityProvider, w.TokenVault, idInfoIdentity, id, info)
+	case driver.IssuerRole:
+		newWallet = newIssuerWallet(w.IdentityProvider, w.TokenVault, id, idInfoIdentity)
+	case driver.AuditorRole:
+		newWallet = newAuditorWallet(w.IdentityProvider, id, idInfoIdentity)
```

Comment:
not sure if it makes sense, but maybe for the sake of completion/clarity I would add a case for driver.CertifierRole and panic because it's not supported. Because it's not clear to everyone why it's left out.